### ISSUE TITLE
OCIS 5.0.3 denies logins with external Keycloak instance

### ISSUE THREAD

**Issue by Kreeblah** [owncloud/ocis#9169]
## Describe the bug
I have an external Keycloak instance that I've been using for a while and works for other systems, which I'd like to use for my OCIS installation.  For the OCIS install, I'm using the plain x86_64 Linux binary for (not in a container) and placing it behind an Nginx reverse proxy.  However, whenever I try to log in, I receive the following error:

Not logged in
This could be because of a routine safety log out, or because your account is either inactive or not yet authorized for use. Please try logging in after a while or seek help from your Administrator.

## Steps to reproduce
I created `/etc/ocis/ocis.env` with the following:

```
# Keycloak IDP configuration
PROXY_AUTOPROVISION_ACCOUNTS=true
PROXY_ROLE_ASSIGNMENT_DRIVER=oidc
OCIS_OIDC_ISSUER=https://keycloak.mydomain.com/realms/myrealm
WEB_OIDC_AUTHORITY=https://keycloak.mydomain.com/realms/myrealm
PROXY_OIDC_REWRITE_WELLKNOWN=true
WEB_OIDC_CLIENT_ID=myclientid
WEB_OIDC_METADATA_URL=https://keycloak.mydomain.com/realms/myrealm/.well-known/openid-configuration

# General config
OCIS_URL=https://ocis.mydomain.com
OCIS_LOG_LEVEL=warn
PROXY_HTTP_ADDR=0.0.0.0:9200
PROXY_TLS=false
PROXY_USER_OIDC_CLAIM=preferred_username
PROXY_USER_CS3_CLAIM=username
OCIS_INSECURE=false
OCIS_ADMIN_USER_ID=
OCIS_EXCLUDE_RUN_SERVICES=idp
GRAPH_ASSIGN_DEFAULT_USER_ROLE=false
GRAPH_USERNAME_MATCH=none

OCIS_CONFIG_DIR=/etc/ocis
OCIS_BASE_DATA_PATH=/var/lib/ocis
```

My nginx config file is:

```
server {
        listen 80 ;
        listen [::]:80 ;

        server_name ocis ocis.mydomain.com;
        return 301 https://ocis.mydomain.com$request_uri;
}

server {
        # default 443 but can deviate if set in OCIS_URL
        listen 443 ssl http2;
        listen [::]:443 ssl http2;

        server_name ocis ocis.mydomain.com;

        ssl_certificate /etc/nginx/certs/ocis.mydomain.com-fullchain.pem;
        ssl_certificate_key /etc/nginx/certs/ocis.mydomain.com-key.pem;
	ssl_dhparam /etc/nginx/dhparam.pem;
	ssl_ecdh_curve secp521r1:secp384r1;

        # restrict ciphers
        #
        # IMPORTANT: depending on your setup, it is possible that certbot will also define the ciphers used.
        # this can lead to a nginx error about already defined ciphers. in such a case, you can comment out
        # the cipher definition here.
        ssl_ciphers "ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-RSA-AES256-SHA:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:!SHA1:!SHA256:!SHA384:!RC4:!aNULL:!eNULL:!Medium:!LOW:!3DES:!MD5:!EXP:!PSK:!SRP:!DSS:!SEED";

        # reduce time to first byte
        ssl_buffer_size 4k;

        # add strict transport security
        add_header Strict-Transport-Security "max-age=15768000; must-revalidate; includeSubDomains; preload;" always;

        location / {
            # OIDC Tokens in headers are quite large and can exceed default limits of reverse proxies
            proxy_buffers 4 256k;
            proxy_buffer_size 128k;
            proxy_busy_buffers_size 256k;

            # Disable checking of client request body size
            client_max_body_size 0;

            proxy_pass http://ocis.ipaddress.here:9200;
            proxy_set_header Host $host;
        }
}
```

I also have this starting up via systemd:

```
[Unit]
Description=OCIS server

[Service]
Type=simple
User=ocis
Group=ocis
EnvironmentFile=/etc/ocis/ocis.env
ExecStart=/usr/local/bin/ocis server
Restart=always

[Install]
```

I also created the YAML file with:

`sudo -u ocis ocis init --config-path /etc/ocis`

And, I'm starting the service with systemd.  Once it's started, I'm going to `https://ocis.mydomain.com` which sends me to the realm login page on my Keycloak instance, I authenticate there, and I'm passed back to the OCIS domain, where I receive the error I mentioned.

## Expected behavior
I'd expect to be able to be able to log in.

## Actual behavior
I'm receiving that error there.

## Setup
See above.

## Additional context
I'm wondering whether there's something different I need to do with how I have the client configured in Keycloak.  Currently, I have it set up with just the standard flow, but maybe there's something else I need to do?

**Comment by Kreeblah** [comment]
So, after figuring out how to get logging to a file going, I found this:

```
{"level":"error","service":"proxy","error":"Get \"https://ocis.domain.com/.well-known/openid-configuration\": EOF","middleware":"oidc wellknown rewrite","url":"https://ocis.domain.com/.well-known/openid-configuration","time":"2024-05-14T20:59:05-07:00","message":"get information from url failed"}
{"level":"error","service":"proxy","error":"failed to verify access token: 502 Bad Gateway: <html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body>\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.22.1</center>\r\n</body>\r\n</html>\r\n","authenticator":"oidc","path":"/ocs/v2.php/apps/notifications/api/v1/notifications/sse","time":"2024-05-14T20:59:06-07:00","message":"failed to authenticate the request"}
```

I don't know whether that's relevant, but it keeps getting written to the logs over and over.  I'd have thought that the well-known URL would have been taken care of via the `PROXY_OIDC_REWRITE_WELLKNOWN=true` I had.

I also found the sample Keycloak client configs at https://github.com/owncloud/ocis/tree/master/deployments/examples/ocis_keycloak/config/keycloak/clients, so I imported those into my Keycloak instance (changing the client IDs and secrets) to hopefully rule out any Keycloak configuration issues.

**Comment by micbar** [comment]
Ocis cannot connect to the well known endpoint from keycloak. Make sure that ocis can connect to the external keycloak hostname.

**Comment by Kreeblah** [comment]
Copying the URL from what I have set for `WEB_OIDC_METADATA_URL` in the .env file (to make sure there aren't any typos), I can curl it from the host:

```
root@hostname:/etc/ocis# curl 'https://keycloak.domain.com/realms/myrealm/.well-known/openid-configuration'
{"issuer":"https://keycloak.domain.com/realms/myrealm","authorization_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/auth","token_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/token","introspection_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/token/introspect","userinfo_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/userinfo","end_session_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/logout","frontchannel_logout_session_supported":true,"frontchannel_logout_supported":true,"jwks_uri":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/certs","check_session_iframe":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/login-status-iframe.html","grant_types_supported":["authorization_code","implicit","refresh_token","password","client_credentials","urn:openid:params:grant-type:ciba","urn:ietf:params:oauth:grant-type:device_code"],"acr_values_supported":["0","1"],"response_types_supported":["code","none","id_token","token","id_token token","code id_token","code token","code id_token token"],"subject_types_supported":["public","pairwise"],"id_token_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"id_token_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"id_token_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"userinfo_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512","none"],"userinfo_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"userinfo_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"request_object_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512","none"],"request_object_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"request_object_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"response_modes_supported":["query","fragment","form_post","query.jwt","fragment.jwt","form_post.jwt","jwt"],"registration_endpoint":"https://keycloak.domain.com/realms/myrealm/clients-registrations/openid-connect","token_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"token_endpoint_auth_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"introspection_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"introspection_endpoint_auth_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"authorization_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"authorization_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"authorization_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"claims_supported":["aud","sub","iss","auth_time","name","given_name","family_name","preferred_username","email","acr"],"claim_types_supported":["normal"],"claims_parameter_supported":true,"scopes_supported":["openid","profile","email","address","roles","acr","web-origins","microprofile-jwt","phone","offline_access"],"request_parameter_supported":true,"request_uri_parameter_supported":true,"require_request_uri_registration":true,"code_challenge_methods_supported":["plain","S256"],"tls_client_certificate_bound_access_tokens":true,"revocation_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/revoke","revocation_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"revocation_endpoint_auth_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"backchannel_logout_supported":true,"backchannel_logout_session_supported":true,"device_authorization_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/auth/device","backchannel_token_delivery_modes_supported":["poll","ping"],"backchannel_authentication_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/ext/ciba/auth","backchannel_authentication_request_signing_alg_values_supported":["PS384","RS384","EdDSA","ES384","ES256","RS256","ES512","PS256","PS512","RS512"],"require_pushed_authorization_requests":false,"pushed_authorization_request_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/ext/par/request","mtls_endpoint_aliases":{"token_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/token","revocation_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/revoke","introspection_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/token/introspect","device_authorization_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/auth/device","registration_endpoint":"https://keycloak.domain.com/realms/myrealm/clients-registrations/openid-connect","userinfo_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/userinfo","pushed_authorization_request_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/ext/par/request","backchannel_authentication_endpoint":"https://keycloak.domain.com/realms/myrealm/protocol/openid-connect/ext/ciba/auth"},"authorization_response_iss_parameter_supported":true}
```

I don't have anything preventing the OCIS binary from making outbound calls, so it should be able to reach the same host.

**Comment by micbar** [comment]
According to the logfile the web client tries to connect to your ocis domain instead of the keycloak domain for the well known endpoint.

**Comment by Kreeblah** [comment]
Yep, which is what I found strange when I noticed that in the log.  And I do get forwarded over in the browser to my Keycloak login page, but when I do, I get another request to the well-known endpoint on the OCIS FQDN:

```
{"level":"error","service":"proxy","error":"failed to verify access token: Get \"https://ocis.domain.com/.well-known/openid-configuration\": context canceled","authenticator":"oidc","path":"/ocs/v1.php/cloud/user","time":"2024-05-14T22:32:00-07:00","message":"failed to authenticate the request"}
```

Completing the login workflow on the Keycloak side then gets me this in the OCIS logs:

```
{"level":"warn","service":"frontend","pkg":"rhttp","traceid":"fe5ad98a0f1e275800aded0ed8497851","time":"2024-05-14T22:34:44-07:00","message":"core access token not set"}
{"level":"error","service":"proxy","error":"Get \"https://ocis.domain.com/.well-known/openid-configuration\": EOF","middleware":"oidc wellknown rewrite","url":"https://ocis.domain.com/.well-known/openid-configuration","time":"2024-05-14T22:34:52-07:00","message":"get information from url failed"}
{"level":"error","service":"proxy","error":"failed to verify access token: 502 Bad Gateway: <html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body>\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.22.1</center>\r\n</body>\r\n</html>\r\n","authenticator":"oidc","path":"/ocs/v1.php/cloud/user","time":"2024-05-14T22:34:54-07:00","message":"failed to authenticate the request"}
```

As far as I can tell, what I have in the .env file between the `WEB_OIDC_METADATA_URL` and `PROXY_OIDC_REWRITE_WELLKNOWN` values there should be rewriting requests to use the Keycloak FQDN for the well-known endpoint, with `OCIS_OIDC_ISSUER` and `WEB_OIDC_AUTHORITY` taking care of everything else.  But, since it doesn't seem to be doing that, I'm at a loss as to where I'm going wrong, since I thought I had everything set to direct everything IDP-related to the Keycloak FQDN.

**Comment by micbar** [comment]
@rhafer @butonic 
Do you have any ideas?

**Comment by rhafer** [comment]
I don't have a concrete hunch yet. What is weird is that there actually is a client trying to access the 'well-known' endpoint via the ocis URL at all. AFAICT that should not happen. 

@Kreeblah could you please increase the `proxy` log-level to `debug` (just set `PROXY_LOG_LEVEL=debug` in your env file)  and upload the logs? Maybe that gives a  hint of what's going on.

Also some notes:
I don't think you need to set `WEB_OIDC_METADATA_URL` and `WEB_OIDC_AUTHORITY` at all. With `OCIS_OIDC_ISSUER` set  correctly ocis will automatically set the right values for the  `WEB_OIDC_*` bits.  Also I think it should be possible to run this setup without PROXY_OIDC_REWRITE_WELLKNOWN=true.

**Comment by Kreeblah** [comment]
I am . . . thoroughly confused here.  You're right that I didn't need to set `WEB_OIDC_METADATA_URL` or `WEB_OIDC_AUTHORITY` at all, nor did I need to set `PROXY_OIDC_REWRITE_WELLKNOWN=true`.  Removing those got me a different error about not finding a role in the claim, so I created a client scope in Keycloak to list out each user's assigned roles at the top level, and things started working.

However, switching back to my previous configuration file in order to try to track down what was causing the original issue I was running into, even with removing all of the data from the installation to mimic a fresh install and removing the client scope in Keycloak (not that that should have made a difference if OCIS wasn't calling out to Keycloak in the first place), I can't reproduce it any longer.

So, I'm up and running, but I'm mystified as to why things weren't working before.  Other than needing to assign a client scope to the OCIS client (which gave me a different message in the logs than I was getting previously), I'm not doing anything differently.

In case anybody else happens to run into this somehow, here's what I ended up with for my final configuration:

```
# Keycloak IDP Configuration
PROXY_AUTOPROVISION_ACCOUNTS=true
PROXY_ROLE_ASSIGNMENT_DRIVER=oidc
OCIS_OIDC_ISSUER=https://keycloak.domain.com/realms/myrealm
WEB_OIDC_CLIENT_ID=keycloak_client_id

# OCIS Configuration
OCIS_URL=https://ocis.domain.com
OCIS_LOG_LEVEL=warn
PROXY_HTTP_ADDR=0.0.0.0:9200
PROXY_TLS=false
PROXY_USER_OIDC_CLAIM=preferred_username
PROXY_USER_CS3_CLAIM=username
PROXY_ENABLE_BASIC_AUTH=true
OCIS_INSECURE=false
OCIS_ADMIN_USER_ID=
OCIS_EXCLUDE_RUN_SERVICES=idp
GRAPH_ASSIGN_DEFAULT_USER_ROLE=false
GRAPH_USERNAME_MATCH=none

OCIS_CONFIG_DIR=/etc/ocis
OCIS_BASE_DATA_PATH=/var/lib/ocis
```

**Comment by micbar** [comment]
@Kreeblah What is your difference to our public keycloak example?

In the example, we showcase how to set up ocis with keycloak in autoprovisioning mode with the roles coming from OIDC an OIDC claim.

**Comment by Kreeblah** [comment]
I haven't analyzed your Keycloak instance configuration super thoroughly, but I imported the client JSON payloads you had and then created the roles for autoprovisioning.  I created a writeup [here](https://github.com/Kreeblah/ocis_config) with info about how to get this going based on my end state, both for my own benefit in case I ever need to get this set up again, and in case anybody else comes across it.

When this wasn't working, I hadn't had the client scope set up, though I did have the roles created at the realm level.  That didn't really seem to matter, though, since OCIS wasn't reaching out to Keycloak for authentication, and I'm still not sure what the difference is.  Removing all my data and setting it up fresh, just as I had it, doesn't get me the same behavior I was seeing before, despite that I'd tried removing my data and setting it up fresh multiple times prior to filing this issue.
